rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,108 @@
+
+Further, for our Cloud offering, we don't perform account resets as a policy. The only recourse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data loss.
+
+On-prem customers could work around this "locked out" state by running `tctl` locally on the auth server to create a new admin user.
+
+## Details
+
+> Context: SSO users
+>
+> Teleport has this feature called SSO auth. It allows for the addition of an SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
+>
+> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
+> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
+> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
+> 4. SSO users can have full access.
+
+The solution here could be to introduce a new rule:
+
+- There should be at least _admin_
+
+This will ensure that:
+
+- Users won't be able to delete _admin_ user if there is only one _admin_. They would have to add a second one and then they can delete it.
+- Users won't be able to unassign the user/role editing capabilities from _admin_ user if there are only is only one _admin_. They would have to add a second one and then they can unassign.
+- Locking will be impossible on the SSO side since there will be always _admin_ on our side.
+
+Now we need to define what _admin_ means in our system. W can take 2 approaches:
+
+1. User with an immutable `editor` role is _admin_.
+2. User with a role that grants user/role editing capabilities is _admin_.
+
+From the user's perspective, the first approach is simpler. The error message saying: "You can't delete the last user with `editor` role." will be easier to understand than "You can't delete the last user with: list of role management capabilities".
+
+The other benefit is that we wouldn't need to define validation for editing roles (since the `editor` role would be immutable).
+
+The drawbacks here are:
+
+- we force users to have the `editor` role
+- complicated migration in case of updating cluster. It is still possible to delete/edit default roles so not all existing clusters will have default roles in the initial form. Maybe the `editor` role is used as a role with a different meaning than the initial one? We can't override existing roles. So the solution would be to add 3 new roles (instead of editing/re-adding: editor, auditor, access). This will ensure we don't break anything. But the problem will be users would have to assign manually a new _admin_ role to some users.
+
+I think `2.` approach is better since migration cost outweighs the benefits of the pros.
```

Comment:
In practice it seems like option 2. is the same as having a fixed root user in cluster, but with some extra complexity added.

Reply:
The main difference here is customer impact:
* (fixed root user) Impacts all clusters. Will require manual migration: adding new root user.
* (option 2)  Most existing clusters have users who can edit roles, so there will be no required action. The only exception is when there are only SSO users in the cluster. We can warn them that adding such a user with x role is advisable (after that they cannot go back to only SSO users state).

Reply:
> (fixed root user) Impacts all clusters. Will require manual migration: adding new root user.

Migration, yes, but why does it have to be manual? I don't follow that.

> (option 2) Most existing clusters have users who can edit roles, so there will be no required action (...)

Automatically adding a root-like user solves this too and requires no action on part of users. I do think we should explore this alternative a bit more before discarding it.

Reply:
Right, we could add a root user automatically using username and password generation (running some procedure when a cluster is updated or initially bootstrapped). 

The authorized user can then reset authentication and set a known password or another form of auth. This is optional but required for the root account to be useful. We could add the warning that it is recommended action.

I think we should consider this: 
I think customers will mostly use passwords as auth since the root user will be not a personal account. Shared passwords are much more likely to be misused, their passwords tend to remain unchanged for extended periods, and often leak when employees change jobs. Also, since they are nobody's personal responsibility and sort fo common knowledge among peers, they tend to not get the same amount of diligence as personal accounts and are often emailed or written down in notes, files, and password managers. Hackers could utilize that to gain access to the cluster.